### PR TITLE
Force customers to fill plate samples correctly in form

### PR DESCRIPTION
**Pull request by elevu**
This PR limits the possibility of submitting samples in 96 well plates when the layout is not column ordered

**How to prepare for test**:
- [x] ssh to clinical-db
- [x] install on stage of clinical-db:
`bash servers/resources/clinical-db.sclifelab.se/update-cgweb-stage.sh [THIS-BRANCH-NAME]`

**How to test**:
- [x] login on https://clinical-stage.scilifelab.se/
- [x] try uploading different types of samples to find 

**Expected test outcome**:
- [x] check that the validation works as expected
- [x] Take a screenshot and attach or copy/paste the output.

Screen Shot 2020-10-29 at 10 45 40

**Review:**
- [ ] code approved by
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by elevu** [comment]
I tested this branch extensively but the form validation logic is really complex, I would appreciate if @annaengstrom or @linnealafleur could test a bit since you might know better how customers use the form. This branch is deployed on staging so you can play with it here: https://clinical-stage.scilifelab.se/order

File: store/getters.js [hunk 1]
```diff
@@ -627,6 +630,55 @@ function exists(value, list) {
...
 }
 
+// If samples are sent in a 96 wells plate, this function returns false if the lay out is not column ordered
+function isColumnLayout(samples, currentSample, wellsLayout) {
+  let isLocalColumnLayout = true;
+  let failingPlates = [];
+  let plates = {};
+  try {
+    samples.forEach((sample) => {
+      if (sample.container === "96 well plate" && sample.containerName) {
```
Comment (patrikgrenfeldt) [inline store/getters.js:640]: Isn't there a function for checking if it is a well-plate? Maybe we could get rid of the hard-coded  "96 well plate" here?

**Review by patrikgrenfeldt — APPROVED** [review]
Awesome! Great that you asked the our lab colleagues to test as well since I'm not sure what column-ordered means, I thought it meant to fill up all columns in one row before starting on the next row 😆

File: store/getters.js (hunk 1 above)
Comment (elevu) [reply store/getters.js]: Oh, found it!

**Comment by annaengstrom** [comment]
Amazing job @elevu! 

I made a first try with this [**orderform**](https://github.com/Clinical-Genomics/cgweb/files/5460125/MWL_more.than88-.2020-10-29-1603-9_1603-9_Order_form_Microbial_WGS.xlsx) for Microbial samples, and I got the very nice picture of what is wrong, I did not place any samples in column 12
Screenshot 2020-10-29 at 18 09 03.


.
Our sample requirements for xxxNXTxxxx apptags is that our customers cannot place any samples in column 12. See **project/ticket 385752** for example
Screenshot 2020-10-29 at 18 16 15.

What is your thought on this @cesve?

**Comment by annaengstrom** [comment]
I also checked, for *Clinical samples* orderform:

- 100 samples, placed in "Plate1" but all in correct position - FAILURE (correct)
- 100 samples, placed in "Plate1 and Plate2" but all in correct position - PASS (correct)
- 98 samples, placed in "Plate1 and Plate2" but with empty wells - FAILURE (correct)
- 4 samples, placed in "Plate1 and Plate2" but with empty wells - **PASS (incorrect)**


[98samples_differentPlate](https://github.com/Clinical-Genomics/cgweb/files/5460280/98samples_differentPlate_1508-21_1508_21_Orderform_Clinical_samples.xlsx)
[100samples_differentPlate](https://github.com/Clinical-Genomics/cgweb/files/5460282/100samples_differentPlate_1508-21_1508_21_Orderform_Clinical_samples.xlsx)
[100samples_samePlate](https://github.com/Clinical-Genomics/cgweb/files/5460283/100samples_samePlate_1508-21_1508_21_Orderform_Clinical_samples.xlsx)

This one passed, I am not sure why:
[4samples_differentPlate](https://github.com/Clinical-Genomics/cgweb/files/5460308/4samples_differentPlate_1508-21_1508_21_Orderform_Clinical_samples.xlsx)

**Review by annaengstrom — CHANGES_REQUESTED** [review]
Good job @elevu, but it seems changes are necessary, see my comments above. Ask if anything is unclear! I can also do more tests if needed.

**Comment by elevu** [comment]
Thank you for testing so well @annaengstrom ! 😃 

> Our sample requirements for xxxNXTxxxx apptags is that our customers cannot place any samples in column 12. See **project/ticket 385752** for example

My understanding after we spoke yesterday with @cesve is that we would wait with enforcing the rule in the form of the 12th column since there was an ongoing discussion. Should I enforce it?

> 4 samples, placed in "Plate1 and Plate2" but with empty wells - PASS (incorrect) A1: B:1 A1: B:1

Why is this incorrect? Must all the wells in a plate be filled before having a second plate?

**Comment by annaengstrom** [comment]
@elevu yes all plate positions in "plate 1" need to be filled before "plate 2" can be filled, the only exception is NXT app tags, where we do not allow samples in column 12 (at least for the coming year).

I did some more testing and consecutive gaps in Plate 1 works (should be FAIL), but random gaps does not work, see https://docs.google.com/spreadsheets/d/1iRflJ6JdIlkhoQhy0MevTFbJVQIs5wE_3WABKkisjzk/edit#gid=0

**Comment by elevu** [comment]
> @elevu yes all plate positions in "plate 1" need to be filled before "plate 2" can be filled, the only exception is NXT app tags, where we do not allow samples in column 12 (at least for the coming year).

Great, then I will apply this rule! I think we should add this information in the Provkrav page as well, maybe it is there but I missed it! There is a "Do not leave any empty wells in between samples." but the way I understood it is that it was in a single plate.


> I did some more testing and consecutive gaps in Plate 1 works (should be FAIL), but random gaps does not work, see https://docs.google.com/spreadsheets/d/1iRflJ6JdIlkhoQhy0MevTFbJVQIs5wE_3WABKkisjzk/edit#gid=0

As it is now no rule is applied about the first plate needing to be filled before using a second plate. So if these gaps come after the samples and the samples that's why it is not working. I will apply this rule... but if the gaps come in the plate before the samples, then it is a bug. Since I cannot replicate this bug, I get FAILED for consecutive gaps before the samples, would you mind sending me the excel file you are used?

Thank you :D

**Comment by annaengstrom** [comment]
@elevu Orderforms for testing are [here](https://drive.google.com/drive/u/0/folders/11JwqKgdaXoZE9X_A8uoVzF7YLj0Q7DGG).

**Description on website for all samples except NXT app tags:**
*Load the samples into the plate column wise and not row wise. Sample 1 should be in well A1, sample 2 in well B1 etc. Note Do not leave any empty wells in between samples.*

**Description on website for NXT app tags:**
*Load the samples into the plate column wise and not row wise. Sample 1 should be in well A1, sample 2 in well B1 etc. Note: Leave column 12, A12-H12, empty. This column is used for internal controls.*

I suppose this could be even more clear. I think the main problem is that very few customers actually read this text. Implementing a "I have read and understood the sample requirements" button would be great.

**Comment by cesve** [comment]
> Thank you for testing so well @annaengstrom ! 😃
> 
> > Our sample requirements for xxxNXTxxxx apptags is that our customers cannot place any samples in column 12. See **project/ticket 385752** for example
> 
> My understanding after we spoke yesterday with @cesve is that we would wait with enforcing the rule in the form of the 12th column since there was an ongoing discussion. Should I enforce it?
> 
> > 4 samples, placed in "Plate1 and Plate2" but with empty wells - PASS (incorrect) A1: B:1 A1: B:1
> 
> Why is this incorrect? Must all the wells in a plate be filled before having a second plate?

Great work @elevu , this will help us a lot! 
Regarding the xxxNXTxxx app tag. Yes I think we should wait with enforcing the "leave column 12 empty". @annaengstrom I think we should have the same requirements for all applications, either we decide to leave column 12 empty for all applications or not. The reason for not having it for the other applications is because we measured the conc. on Qubit before. What do you think, should we accept full plates (except 1 well for internal controls? It will be more efficient in the prep.

**Comment by annaengstrom** [comment]
> > Thank you for testing so well @annaengstrom ! 😃
> > > Our sample requirements for xxxNXTxxxx apptags is that our customers cannot place any samples in column 12. See **project/ticket 385752** for example
> > 
> > 
> > My understanding after we spoke yesterday with @cesve is that we would wait with enforcing the rule in the form of the 12th column since there was an ongoing discussion. Should I enforce it?
> > > 4 samples, placed in "Plate1 and Plate2" but with empty wells - PASS (incorrect) A1: B:1 A1: B:1
> > 
> > 
> > Why is this incorrect? Must all the wells in a plate be filled before having a second plate?
> 
> Great work @elevu , this will help us a lot!
> Regarding the xxxNXTxxx app tag. Yes I think we should wait with enforcing the "leave column 12 empty". @annaengstrom I think we should have the same requirements for all applications, either we decide to leave column 12 empty for all applications or not. The reason for not having it for the other applications is because we measured the conc. on Qubit before. What do you think, should we accept full plates (except 1 well for internal controls? It will be more efficient in the prep.

Can we not have different enforcements for different app tags @elevu @cesve?

Manually it will be complicated to prepare micro samples with samples in column 12, and for all other applications it is good to fill the entire plate.

**Comment by cesve** [comment]
> > > Thank you for testing so well @annaengstrom ! 😃
> > > > Our sample requirements for xxxNXTxxxx apptags is that our customers cannot place any samples in column 12. See **project/ticket 385752** for example
> > > 
> > > 
> > > My understanding after we spoke yesterday with @cesve is that we would wait with enforcing the rule in the form of the 12th column since there was an ongoing discussion. Should I enforce it?
> > > > 4 samples, placed in "Plate1 and Plate2" but with empty wells - PASS (incorrect) A1: B:1 A1: B:1
> > > 
> > > 
> > > Why is this incorrect? Must all the wells in a plate be filled before having a second plate?
> > 
> > 
> > Great work @elevu , this will help us a lot!
> > Regarding the xxxNXTxxx app tag. Yes I think we should wait with enforcing the "leave column 12 empty". @annaengstrom I think we should have the same requirements for all applications, either we decide to leave column 12 empty for all applications or not. The reason for not having it for the other applications is because we measured the conc. on Qubit before. What do you think, should we accept full plates (except 1 well for internal controls? It will be more efficient in the prep.
> 
> Can we not have different enforcements for different app tags @elevu @cesve?
> 
> Manually it will be complicated to prepare micro samples with samples in column 12, and for all other applications it is good to fill the entire plate.

I think it's possible to have different requirements for different applications. But do we want that? I think it will be easier also in the micro-prep to prepare full plates? Especially when transferring to 384-plates. And we also need an internal control in the TGA-prep.

**Comment by annaengstrom** [comment]
@cesve maybe we need to discuss this in a separate meeting?

**Comment by annaengstrom** [comment]
Decided 2020-11-02 between @cesve and @annaengstrom:

- xxxNXTxxxRxxx: always leave column 12 empty
- all other app tags: only full plates before next plate is filled in

**Comment by elevu** [comment]
@annaengstrom great, I will proceed adding those requirements!

**Comment by elevu** [comment]
> * xxxNXTxxxRxxx: always leave column 12 empty

Do you mean xxxNXTRxxxx ?
Or xxxNXTxxxx?

Could not find any application with pattern xxxNXTxxxRxxx

**Comment by cesve** [comment]
> > * xxxNXTxxxRxxx: always leave column 12 empty
> 
> Do you mean xxxNXTRxxxx ?
> Or xxxNXTxxxx?
> 
> Could not find any application with pattern xxxNXTxxxRxxx

It should be xxxNXTRxxx @elevu !

**Comment by elevu** [comment]
Opened another PR https://github.com/Clinical-Genomics/cgweb/pull/236

**Pull request closed without merging**